### DISCUSSION TITLE
Refaktorisanje na ViewModele

### DISCUSSION THREAD

**Discussion by igr** [igr/Clmn#6] [General]
@vexy Pišem ovde da te ne bih smarao na wapp :)

Model ovde je hijerarhija: `Board` -> `TaskList` -> `Task`.

Razumem da treba da napravim e.g. `BoardViewModel`:

```swift
class BoardViewModel: ObservableObject {
    @Published private(set) var boards: [Board] = []

    func loadBoards() {
        boards = [] /// OVDE IH UČITAVAM IZ Servisa/Fridge
    }

    func newBoard(_ name: String) -> Board {...}
    func deleteBoard(_ boardToDelete: Board) -> Board? {... }
}
```

Dakle, ovde prebacim svu biznis logiku vezanu za kreiranje boardova. Takođe, održavam `boards` aktualnim - dodajem u njega i uklanjam elmente, što trigeruje iscrtavanje.

## Pitanja

Šta radim sa relacijama? Ja bih napravio `TaskListViewModel` na sličan način, koji takođe ima listu `[TaskList]` i koja se isto učivata na `onAppear`? Ili to ne treba da radim, jer smo dovukli celo stablo malo pre?

Drugim rečima, da li `Board` treba da ima kolekciju `[TaskList]` u sebi? Da li u Fridge snimam samo Boards (top model) ili snimam nezavisno e.g. `Boards` i `TaskList` e.g. zasebne kolekcije, kako bih to već radio u Mongo.

Takođe, sada na promeni `boards` (kada se dodaje/oduzima element) radim eksplicitno snimanje u Fridge?

**Comment by vexy** [comment]
Prvo - da, većina stvari koje si napisao je neka generalna struktura koja je ovde prikladna. 👍 

Idemo redom, delim pitanje na dva dela:

1. Elementi/Entiteti/Struktura
2. Odgovori na pitanja

### Struktura
Predstaviću ovde više layer-a koje su u igri, a oko kojih možemo da variramo i dalje optimizujemo:
1. modeli
2. "početak aplikacije i uvezivanje"
3. ViewModel/Servis

#### Modeli
`Board` -> `TaskList` -> `Task` će da završe kao neki `struct`, a na koji ćemo kasnije da dodamo `Codable`. Ovako nekako: (pseudo totalno).
```Swift
struct Board {
    // nesto nesto
    var lists: [TaskList]
}

struct TaskList {
   // nest nesto
  var tasks: [Task]
}

struct Task: Codable {  // Codable = protokol za transcoding (kao neki ORM)
  // nesto nesto
}
```

Same detalji sada i nisu bitni...

#### Početak
OS svet počinje od `@main` funkcije. Obično neki struct koji implementira`App` protokol. SwiftUI ima i drugačije načine, ali to sad preskačemo. Ovako to nekako izgleda:

```Swift
@main
struct Clmn: App {
   @StateObject var boardModel: BoardViewModel   //"veza" i desavanje init()-a
   
  // potpuni ignore sto ovo ne prolazi compile sad i sto treba drugacije
  var body: some Scene {
     WindowGroup {
        List(boardModel.allBoards) { //nesto }
        // ili npr:
        ForEach(board) { // nesto nesto } 
     }
  }
}
```

Na ovaj način počne _SwiftUI redraw cycle_ - sam od sebe. Dakle ne mora čak ni `onAppear`.
Medjutim, potrebno je dodati i poslednju stvar a to je `ViewModel` koji si predstavio gore.

#### ViewModel / Servis
```Swift
class BoardViewModel: Observable {  //1.
   @Published private(set) var allBoards: [Board] //2.

   init() {
      allBoards = [] //3. **
      // ovde moze load sa storage-a 
   }
   
   func newBoard(newBoard: Board) {
      allBoards.append(newBoard)   //3. **
   }

   func deleteBoard() { allBoards.remove(xyz) }  //3. **
   func whateverWeNeed() { for b in allBoards {...} }
} 

// takodje ignore sto ne prolazi compile
```
Redom:
1. Observable protokol + `class`
2. @Published property - prosta, obicna varijabla koja je "niz modela"
3. "alteracija stanja" @published propertija pod vodjstvom biznis logike :)

Sve "alteracije stanja" ove varijable (ili bilo kog @published var-a) **automatski** pokreću "redraw cycle". Promene vrednosti se preslikavaju na `Clmn` UI strukturi i ostalim delovim aplikacije koje prate ovaj pattern ("veza").

Iza svega ovoga je u stvari `Combine` framework. On je tipičan `pub/sub` mehanizam u stvari. Published na to ukazuje, a subskrajberi su u stvari view strukture (UI). Glavni štos je što se to i ne vidi na površini 😄 

Ovde stajem sa strukturom, ako ima pitanja pucaj 😄 

# Odgovori
> Šta radim sa relacijama? Ja bih napravio TaskListViewModel na sličan način, koji takođe ima listu [TaskList] i koja se isto učivata na onAppear? Ili to ne treba da radim, jer smo dovukli celo stablo malo pre?

Kako god misliš da je prikladnije. Da, kada se stvari ovako postave, imamo celo stablo odmah. Pattern je generalni tako da otvara vrata za TaskList na istu foru.

> Drugim rečima, da li Board treba da ima kolekciju [TaskList] u sebi?

Može baš tako :) A može i drugačije (vidi ispod).

> Da li u Fridge snimam samo Boards (top model) ili snimam nezavisno e.g. Boards i TaskList e.g. zasebne kolekcije, kako bih to već radio u Mongo.

Bukvalno tako ! Iza Fridge je u stvari `BSON` kao storage format, baš kao u Mongo.
`Identifier` je kao `collection`.
Sam `struct` je kao `document`.

Tako da je ovo vezano za odgovor od pre. Bukvalno svaki Foundation objekat, svaki `struct`, se direktno mapira u neki BSON.
Jedno je samo bitno; da se implementira (doda u stvari) `Codable` na `struct`. Bez toga pljušte error-i :)
Samim tim, sačuvana je tipizacija po želji i potrebi.

> Takođe, sada na promeni boards (kada se dodaje/oduzima element) radim eksplicitno snimanje u Fridge?

E OVO JE ODLIČNO PITANJE 🥇 😄 
Imamo tu XY mogućnosti, ja sam generalno za "reaktivni model", tj da aplikacija _odgovara_ na interupte sistema ili trigere korisnika.
Npr. obzirom da je ovo Mac app, triger za "save/commit" u `Fridge` može biti čak i sam `Quit` aplikacije. Minimize takodje ili neki Save na UI delu koji klikne user.

Do tada predlažem in-memory manipulaciju dok god je moguće. UI dosta manje bude opterećen (milisekunde, ali avaj... godine iskustva :)) a ako se share-uju modeli kroz ostale UI elemente (npr `EvironmentObject` modifierom) komunikacija ostaje dvosmerna i "vidi se" na oba ekrana. Bez potrebe za commit na disk, još uvek.

Takodje, u nekim daljim koracima, npr neki payment confirmation notification - kad stigne, naravno :) - može da pokrene neki lanac čuvanja podataka o tome. Itd...

A naravno, ako misliš da je bolje da se "save" na disk radi ranije - možemo i tako.

@igr

**Reply by igr** [reply]
Jasno, hvala!

Čitao sam malo, ne bih da imam celo drvo u memoriji. TaskList će biti jedno drvo, tj. jedna kolekcija

Cimaću te oko Fridga par pitanja.
